--- a/py-compile
+++ b/py-compile
@@ -116,7 +116,10 @@
 fi
 
 $PYTHON -c "
-import sys, os, py_compile, imp
+import sys, os, py_compile, importlib.util
+from types import SimpleNamespace
+imp = SimpleNamespace(get_tag=sys.implementation.cache_tag,
+                      cache_from_source=importlib.util.cache_from_source)
 
 files = '''$files'''
 
@@ -137,7 +140,10 @@
 
 # this will fail for python < 1.5, but that doesn't matter ...
 $PYTHON -O -c "
-import sys, os, py_compile, imp
+import sys, os, py_compile, importlib.util
+from types import SimpleNamespace
+imp = SimpleNamespace(get_tag=sys.implementation.cache_tag,
+                      cache_from_source=importlib.util.cache_from_source)
 
 # pypy does not use .pyo optimization
 if hasattr(sys, 'pypy_translation_info'):
